mantidproject/mantid
# New shoebox integration algorithm for single-crystal Bragg peaks

### Description of work

New shoebox integration algorithm for single-crystal Bragg peaks - see docs for details.
It's quicker than the skew integration and doesn't require converting the data to Q or HKL.

It has a few advantages over existing algorithms:

- Shoebox dimensions of weak peaks can be taken from nearby strong peaks
- Avoids nearby peaks that are closer to another peak position when optimising shoebox position
- Produces a pdf output like `IntegrtaePeaksSkew`

Fixes #36322 

**Report to:** Pas (WISH)

### To test:

(1) Run this script and inspect the output
```
ws = Load(Filename='SXD23767.raw', OutputWorkspace='SXD23767')
peaks = CreatePeaksWorkspace(InstrumentWorkspace='SXD23767', NumberOfPeaks=0, OutputWorkspace='peaks')
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=8303.3735339704781, DetectorID=7646)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=19166.422281671705, DetectorID=20009, Height=7489, BinCount=2.4956385131770986)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=1582.4728457415549, DetectorID=25111, Height=516, BinCount=13.143872508760753)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=1734.9661945986509, DetectorID=28271, Height=203, BinCount=16.920867383989862)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=5241.7309709929505, DetectorID=1380, Height=7422, BinCount=1513.6455817753372)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=1665.2648421691604, DetectorID=8176, Height=585, BinCount=36.078854751614564)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=7690.5100117616385, DetectorID=22799, Height=489, BinCount=0.79030785093533673)
AddPeak(PeaksWorkspace='peaks', RunWorkspace='SXD23767', TOF=8289.2655006925579, DetectorID=7327, BinCount=0.0018390042716158461)

IntegratePeaksShoeboxTOF(InputWorkspace=ws, PeaksWorkspace=peaks, 
                         OutputWorkspace="peaks_out", GetNBinsFromBackToBackParams=True, 
                         WeakPeakThreshold=0.0, LorentzCorrection=False,
                         OutputFile=r"C:\Users\xhg73778\Desktop\shoebox_test.pdf")
```
(2) Check the pdf output has

- No page for ipk=1 (peak is out of TOF range of data)
- I/sigma=0 for ipk=5 (peak on detector edge)
- Ipk=7 hasn't jumped onto the peak in the data because there is already a  peak in the table that is closer (ipk=0)

(3) Run this with `IntegrateIfOnEdge=True` 
```
IntegratePeaksShoeboxTOF(InputWorkspace=ws, PeaksWorkspace=peaks, 
                         OutputWorkspace="peaks_out_edge", GetNBinsFromBackToBackParams=True, 
                         WeakPeakThreshold=0.0, LorentzCorrection=False, IntegrateIfOnEdge=True, 
                         OutputFile=r"C:\Users\xhg73778\Desktop\shoebox_test_edge.pdf")
```
- Check ipk=5 has been integrated (I/sigma > 0)

(4) Set `OptimiseShoebox=False`
```
IntegratePeaksShoeboxTOF(InputWorkspace=ws, PeaksWorkspace=peaks, 
                         OutputWorkspace="peaks_out_fixed", GetNBinsFromBackToBackParams=True, 
                         WeakPeakThreshold=0.0, LorentzCorrection=False, IntegrateIfOnEdge=True, 
                         OptimiseShoebox=False,
                         OutputFile=r"C:\Users\xhg73778\Desktop\shoebox_test_fixed.pdf")
```
- Check in the pdf output that the shoeboxes are in general smaller and I/sigma also decreased.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
